### ISSUE TITLE
Improve external dependency management of Reaktoro

### ISSUE THREAD

**Issue by allanleal** [reaktoro/reaktoro#29]
This is a proposal to use a C++ dependency manager to simplify the installation of Reaktoro's external dependencies.

Currently, the external dependencies are embedded into the project (some libraries are zipped under `thirdparty` directory, others are available directly, such as Eigen, header only, copied directly to the source directory). CMake command `ExternalProject_Add` is then used to install each dependency.

By using a C++ dependency manager, these external projects need to be build and installed only once per machine, instead of every time Reaktoro is built from scratch.

Checking the available options (conan, hunter, vcpkg, etc), I found conan to be quite interesting, easy to use, and perhaps the most mature C++ dependency manager.

Any thoughts? Any other option to consider?

**Comment by nicoddemus** [comment]
Conan is (IUC) C++ only, correct? I think we also plan to support Python users.

We have been using `conda` for C++ and Python here at ESSS for several years now, with great success. We believe it would be an excellent fit for Reaktoro.

**Comment by allanleal** [comment]
I think `conda` and `Conan` are similar someway, but serve different purposes, but I could be wrong as both things are relatively new to me. 

Conan is a C++ package manager that also integrates well with CMake so that external dependencies can be installed on the fly during (actually before) build time. 

Building any library that uses Conan as an external dependency manager requires a single step before cmake configure step (commands below under build dir):

~~~
conan install ..
cmake ..
make
~~~

Can `conda` manage external dependencies of C++ projects in a similar way? 

The idea here is not to have one instead of the other, but use both technologies in the project if needed. However, if we can achieve more or less the same convenience with `conda` only, then we should stick with it and not introduce Conan as a new dependency.

**Comment by allanleal** [comment]
It turns out many other C++ projects are there in conda-forge already (interesting). I'll keep looking. 

Even if conda has no native CMake integration, we can play along with `ExternalProject_Add` cmake command, for which we define the conda install command for each dep. 

But we should make sure we are not missing anything nice Conan could offer to us before making a final decision.

**Comment by tadeu** [comment]
> The idea here is not to have one instead of the other, but use both technologies in the project if needed. However, if we can achieve more or less the same convenience with `conda` only, then we should stick with it and not introduce Conan as a new dependency.

I think this is possible.

Even though `conda` has no native CMake integration, a lot of C++ conda packages come with cmake "configuration" inside (e.g., `pybind11` comes with `shared/cmake/pybind11/pybind11Config.cmake`), so this integration is almost unnecessary: `conda` installs the dependencies, and `cmake` finds them using `find_package(...)`.

Some Reaktoro dependencies would have to be added or updated in conda-forge (MINIZ, CVODE), but, as you noted, some are already there:

https://github.com/conda-forge/pugixml-feedstock
https://github.com/conda-forge/nlohmann_json-feedstock

Conda-forge is one of the things that makes `conda` more awesome: it's a great infrastructure for building and maintaining binary (or source-only) packages, no matter the language. I don't know if there's a comparable "infrastructure" for `conan`, at the same level of maturity.

`conda-devenv` is a tool that helps work with packages in "develop" mode, so, in order to build `reaktoro` locally, the workflow would be something as simple as this:

```
conda devenv  # <- creates environment with all dependencies already installed
source activate reaktoro  # <- "activates" the environment, putting things on PATH, etc.
cmake ...
make ... # <- builds it
```

The advantage is that it should be possible to do this afterwards:

```
python -c "import reaktoro"
pytest .
```

and whatever is installed from the build process or from the dependencies should be on a local environment (virtualenv-like), without touching the system configuration.

**Comment by allanleal** [comment]
OK. I think we should then start planning the transition from **all-external-dependencies-embedded-in-the-project** to **most-or-all-external-dependencies-installed-using-conda**.

If we classify the dependencies as `internal` and `external`, where internal dependencies are those we develop ourselves, and external are those developed by third-party groups (e.g., eigen, pybind11), then I would say we should rely on Conda to install the external dependencies. For the internal ones, we should rely on git submodule. The reason is that the internal library available as a git submodule can be quickly changed, with immediate effect in Reaktoro. If the internal ones are also installed via Conda, then there will be some delay between changing them and seeing the effects in Reaktoro. Maybe you know a better solution for this.

Note that the above does not imply that these internal libraries should not be uploaded to conda-forge. We'll probably make them all available in conda-forge at some point.

**Comment by allanleal** [comment]
Closing this. Let's focus on using conda-forge. I'll create a new issue to discuss how we evolve from there.
